### ISSUE TITLE
Creating a redis connection keeps the process running indefinitely.

### ISSUE THREAD

**Issue by ramitmittal** [luin/ioredis#1259]
Merely creating a redisClient causes the process to keep on running. Ideally, the process should exit.

Steps to reproduce:
```
const Redis = require("ioredis");

const REDIS_URL = 'redis://localhost:6379';

const redis = new Redis(REDIS_URL);
```

node version - v14.14.0

ioredis version - 4.19.4

**Comment by luin** [comment]
Hi @ramitmittal,

This behavior is intentional. The process will keep running as long as there are active handles (see https://github.com/mafintosh/why-is-node-running). To exit the process, you can simply disconnect to Redis by calling `redis.quit()`.

**Comment by ramitmittal** [comment]
Is there a way to 'unref' redis handles?

**Comment by stavalfi** [comment]
not sure what is `unref`.

in node, all network connections will block the process from exit until they all disconnect.

**Comment by laverdet** [comment]
Here's a workaround for anyone running into this. We use redis primarily as a pubsub service which listens for events eternally for use in a WebSocket server. In this case we only want the HTTP service keeping the process alive.

```ts
import type { Redis } from "ioredis";

export function unrefRedisClient(redisServer: Redis) {
    const erased = redisServer as any;
    erased.connector.connect = function(connect) {
        return async function(this: unknown, ...args: unknown[]) {
            const connection = await connect.apply(this, args);
            connection.unref();
            return connection;
        };
    }(erased.connector.connect);
    return redisServer;
}
```

As the author mentioned in #44 this *will* cause dropped commands in the case there is still work pending. If you are using any of the `BR`, `BL`, or `BZ` commands I definitely do not recommend using this.

**Comment by if4lcon** [comment]
> Here's a workaround for anyone running into this. We use redis primarily as a pubsub service which listens for events eternally for use in a WebSocket server. In this case we only want the HTTP service keeping the process alive.
> 
> ```ts
> import type { Redis } from "ioredis";
> 
> export function unrefRedisClient(redisClient: Redis) {
>     const erased = redisClient as any;
>     erased.connector.connect = function(connect) {
>         return async function(this: unknown, ...args: unknown[]) {
>             const connection = await connect.apply(this, args);
>             connection.unref();
>             return connection;
>         };
>     }(erased.connector.connect);
>     return redisClient;
> }
> ```
> 
> [You have to use `lazyConnect: true`]
> 
> As the author mentioned in #44 this _will_ cause dropped commands in the case there is still work pending. If you are using any of the `BR`, `BL`, or `BZ` commands I definitely do not recommend using this.

can you please explain how to use this, thanks

**Comment by if4lcon** [comment]
can i  just use `process.exit()` ? 

`.quit()` doesn't doing a thing for me

**Comment by Yuval-Peled** [comment]
Hey all, I'm unsure why this issue was closed. This is still relevant for me.
I wanted to be able to shut down my server properly when receiving SIGTERM. 

`redisClient.quit()` was not good for me because if I ran it immediately when getting SIGTERM then my other processes running in the server might still try to use the redisClient after calling `quit()` which would cause errors.

I was looking for a way to shutdown only idle connections and still accept new connections if needed, similar to `node-pg`'s "idleTimeout" setting. This would be the best solution for me, since it would allow unused connections to close, but connections that are in use would still keep the process open. However, I couldn't find a way to implement it.

Luckily, my use case allows me to close connections while running since I only use Redis for distributed locks. If a process is closing, it doesn't need to lock anything. For this case, I can "unref" all Redis connections.

This code seems to work for me:
```typescript
export const redisClient = new IORedis(config.REDIS_CONN_STRING, { lazyConnect: true });
redisClient.on('connect', () => {
    redisClient.stream.unref();
});
```
